### PR TITLE
Add pattern, support reference constraints on primitives, and add number/integer constraints

### PR DESCRIPTION
**Pull request by nolag**

### CODE REVIEW

**Comment by omissis** [comment]
Hey @nolag , thanks for putting out this PR! Can you give me a heads up when it's ready (I see you're still pushing on it) so I can review it? :)

**Comment by nolag** [comment]
> Hey @nolag , thanks for putting out this PR! Can you give me a heads up when it's ready (I see you're still pushing on it) so I can review it? :)

It's ready, I don't have more work planned.  Each commit I made was meant to be independent, I figured I could open one PR and you could review it by commit if you wanted, but at this point, a squash probably makes more sense.

I can make a final commit tomorrow to resolve the merge conflict.  I'm not sure why, but goland insisted on some of those changes and it wouldn't work without them.

**Comment by omissis** [comment]
sweet! can you just fix the conflict in the go.mod? I'll look at it soon as I can!

**Comment by nolag** [comment]
> sweet! can you just fix the conflict in the go.mod? I'll look at it soon as I can!

I realized I have one more commit locally that I need to clean up.  I'll also fix the go mod issue.  Still aiming for today.

**Comment by nolag** [comment]
@omissis FYI, should be good to go soon, I can't reproduce the generation error locally though...  I'll spend a bit of time to debug it, but if you can reproduce that would be helpful...  I think it has to do with the casting floats to ints, I should probably use some epsilon or a better way like math.Round.  I'll give that a try and ping you if I still need help

Sorry for all the force pushes at the end, I had some trouble getting the linter running locally for the tests (even with upstream/main).  

I did find another bug that I didn't fix.  Arrays don't apply rules for their elements.  Ie: String's length in an array and now patterns and number bounds aren't checked.

**I'll need to fix that eventually (or have it fixed) but it'll be a new PR at this point (it's not blocking our use case right now).**

**Comment by nolag** [comment]
@omissis was able to get it consistently passing locally and on CI now.  PR is good to go.

For context, the min int size is so that we can use binary data as a uint8, which we need.  It's a flagged feature so it won't change any existing users :).  It also allows more efficient checks when the number is bound by a set size.

File: main.go [hunk 1]
```diff
@@ -166,6 +168,11 @@ also look for foo.json if --resolve-extension json is provided.`)
 		"Use the schema title as the generated struct name")
 	rootCmd.PersistentFlags().StringSliceVar(&tags, "tags", []string{"json", "yaml", "mapstructure"},
 		`Specify which struct tags to generate. Defaults are json, yaml, mapstructure`)
+	rootCmd.PersistentFlags().BoolVar(
+		&minSizedInts,
+		"min-sized-ints",
+		false,
+		"Uses sized int and unit values based on the min and max values for the field")
```
Comment (omissis) [inline main.go]: is "unit" or "uint" here?

File: README.md [hunk 2]
```diff
@@ -122,15 +122,15 @@ only specific validations remain to be fully implemented.
     * [x] `type` (multiple; **note**: partial support, limited validation)
     * [ ] `const`
   * [ ] Numeric validation (§6.2)
```
Comment (omissis) [inline README.md]:
```suggestion
  * [x] Numeric validation (§6.2)
```

File: tests/data/validation/primitive_defs/primitive_defs.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      "$ref": "#/$defs/minStr"
+    },
+    "myNullableString": {
+      "$ref": "#/$defs/minStr"
+    }
+  },
+  "required": [
+    "myString"
+  ]
+}
```
Comment (omissis) [inline tests/data/validation/primitive_defs/primitive_defs.json]:
```suggestion
}

```

**Review by omissis — COMMENTED** [review]
Impressive PR! besides a couple nitpicks, I am just not sure about the exclusive minimum and maximum: I see they support two implementations, one with numeric exclusive min/max, and one with boolean + an extra value.
From what I see on jsonschema, the numeric one should be the correct one since draft-07, so I take you implemented both for BC reasons? in that case the boolean one should be the one that has the "Old" suffix, right? I'm referring to tests/data/validation/exclusiveMaximum/exclusiveMaximumOld.[json|go]

Thanks!

**Comment by nolag** [comment]
Thanks for reviewing, and good catch on my mix-up with the exclusive min/max.

**Review by omissis — APPROVED** [review]

**Pull request merged**